tests/test_update_ssh_keys.rs

```diff
@@ -0,0 +1,524 @@
+        let user =
+            users::get_user_by_name(&unamestr).ok_or_else(|| format!("failed to find user with name '{}'", unamestr)).unwrap_or(users::User::new(UID_CORE as uid_t, USERNAME_CORE, GID_CORE as gid_t));
+
+        AuthorizedKeys::open(user, true, Some(ssh_dir)).expect(format!("failed to open authorized_keys directory for user '{}'", unamestr).as_str())
+    }
+
+    // A wrapper for adding an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like "valid1", "bad"
+    // testname: key name given as cmdline args, like "one", "two"
+    // is_force: whether to force adding a key via "--add-force"
+    // is_replace: whether to adding a key by replacing an existing one, unless "--no-replace".
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn add_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str, is_force: bool, is_replace: bool, assert_keys: Vec<&str>) -> Result<(), ErrorKind> {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        let res = aks.add_keys(testname, keys.clone(), is_replace, is_force);
+
+        match res {
+            Ok(keys) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Ok(())
+            }
+            Err(err) => {
+                add_key_check_results(ssh_dir, aks, keys, pubkeyname, testname, assert_keys);
+                Err(err.into())
+            }
+        }
+    }
+
+    // A wrapper for deleting an ssh key.
+    //
+    // pubkeyname: name of ssh public key, like valid1, bad
+    // testname: key name given as cmdline args, like "one", "two"
+    // assert_keys: Vec of keys to be asserted after add_keys() succeeded.
+    fn del_one_ssh_key(ssh_dir: &str, pubkeyname: &str, testname: &str) {
+        let mut aks = open_authorized_keys(ssh_dir);
+        let keyfiles = [format!("{}/{}.pub", ssh_dir, pubkeyname)];
+
+        let mut keys = vec![];
+        for keyfile in keyfiles {
+            let file = File::open(&keyfile).expect(format!("failed to open keyfile '{:?}'", keyfile).as_str());
+            keys.append(&mut AuthorizedKeys::read_keys(file).unwrap_or_default());
+        }
+
+        for key in aks.remove_keys(testname) {
+            if let AuthorizedKeyEntry::Valid {
+                ref key,
+            } = key
+            {
+                assert!(key.to_fingerprint_string().contains(&*FINGERPRINTS[pubkeyname]));
+
+                aks.write().expect("failed to update authorized keys directory");
+                aks.sync().expect("failed to update authorized keys");
```

Comment:
This should be done only once at the end after the for loop

Reply:
That is only true for sync, not for write.
Without running write before checking for key files, the whole test does not run at all.
Moved only sync out of the loop.